internal/cortex/frontend/transport/handler.go

```diff
@@ -122,14 +184,35 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	r.Body = http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize)
 	r.Body = io.NopCloser(io.TeeReader(r.Body, &buf))
 
+	// Store query expression and range length for checks
+	queryExpressionNormalized := NormalizeString(r.URL.Query().Get("query"))              //change to correct key for query expression
+	queryExpressionRangeLength, strConvErr := strconv.Atoi(r.URL.Query().Get("duration")) //change to correct key for query time range length
+
+	if strConvErr != nil { //temporary, remove when correct key found for query time range length
+		writeError(w, strConvErr)
+		return
+	}
+
+	// Check if query in cache and whether value exceeds time range length
+	if value := f.lru.Get(queryExpressionNormalized); value != -1 && value >= queryExpressionRangeLength {
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
 	startTime := time.Now()
 	resp, err := f.roundTripper.RoundTrip(r)
 	queryResponseTime := time.Since(startTime)
 
 	if err != nil {
 		writeError(w, err)
+		queryString = f.parseRequestQueryString(r, buf)
+
+		// If error should be cached, store it in cache
+		if CacheableError(resp.StatusCode) {
+			f.lru.Put(queryExpressionNormalized, int(queryResponseTime.Seconds())) //need to find query expression
```

Comment:
Log this important action.
```
level.Info(util_log.WithContext(r.Context(), f.log)).Log(
  "msg", "Cached the query due to a catchable error", 
  "response", resp,
)
```

Reply:
added info logs for both retrieving cached and adding to cache, will add debug logs throughout testing when necessary